chiuinggum/remote-assignments
# remote-week2: add APIs

1. How did you design the user table schema? Why?
    
   <img width="413" alt="5" src="https://github.com/chiuinggum/remote-assignments/assets/135101542/c808bb5f-f78b-4283-b1a2-55734d9ccad8">

    - Primary key (`id`) with auto-increment
    I designated the `id` field as the primary key of the table and set it to auto-increment. Hence, each new user record added to the table will automatically receive a unique identifier
    - Unique key (`email`)
    I designated the `email` field as a unique key. Since the description in assignment 2 says “email address also seen as the account id”, we know that email address can’t be repeated among users.
    - Null value constraints
    To enforce data integrity and prevent empty or null values for essential fields, that is all of the fields, I set the “null” value option to “NO” for all fields.
3. What does the `status code` mean? Is it a must to use?
    
    HTTP response status codes, as described by MDN Web Docs, "indicate whether a specific HTTP request has been successfully completed.” These codes provide a standardized way for web servers to convey the result of a client’s request.

    - `1xx` — informational responses
      - They are used to indicate that the server has received the request and is continuing to process it
    - `2xx` — successful responses
      - They are used to indicate that the request was successfully received, understood, and processed by the server
      - Most common status codes for successful requests
      - `200` — Success with response body sent back
    - `3xx` — redirection messages
      - They are used to inform the client that further action is needed to complete the request
    - `4xx` — client error responses
      - They are used to indicate that there was an error in the client’s request, and the server cannot or will not process it
      - Typically caused by client-side issues, such as invalid requests, authentication problems, or insufficient permissions
      - `400` — sent when there is a client error, `'Request-Date header not exist.'` / `'Missing required fields in the request body.'` / `'Password not strong enough.'` / `'Email not in the right form.'`
      - `409` — “sent when a request conflicts with the current state of the server”, `'Email Already Exists.'`
      - `415` — sent when “the media format of the requested data is not supported by the server, so the server is rejecting the request”, `'Invalid Content-Type header. Only accept application/json.'`
    - `5xx` — server error responses
      - They are used to indicate that the server encountered an error or is otherwise in capable of performing the request
      - `500` — sent when “the server has encountered a situation it does not know how to handle”, `'Error querying the database.'` / `'Error inserting data.'`

    While it’s not strictly mandatory to use status codes in HTTP responses, they are a fundamental and recommended part of the HTTP protocol.
    
4. How and what tools did you use to test your APIs?
    
    In the initial stages, I attempted to use web forms to implement the POST method for the API. However, I encountered challenges when trying to set the request headers, the "Content-Type" and "Request-Date" headers. Unfortunately, I was unable to send these headers as expected using this approach.
    
    Realizing the limitations of the web form approach, I switched to using Postman for API testing.
    
    - For both APIs, first
        - In the “Headers” field, add ‘Content-Type’ header and set its value to `application/json`
        - In the “Pre-request Script” field, write the code
            
            ```jsx
            const currentDate = new Date().toUTCString();
            
            pm.request.headers.add({
                key: 'Request-Date',
                value: currentDate
            });
            ```
            
    - For GET method
        - Paste the URL `http://13.238.167.235:3000/users/id`, where `id` is a number
        - Click the ‘Send’ button
    - For POST method
        - Paste the URL `http://13.238.167.235:3000/users`
        - In the “Body” field, make the data to send, for example
            
            ```json
            {
                "name":"User25",
                "email":"user25@appworks.com",
                "password":"Test_123"
            }
            ```
            
        - Click the ‘Send’ button
    - The response body will appear in the “Body” field down below.
5. How and what tools did you use to make sure the app would not shut down even if you exit the SSH connection?
    
    I use the command `nohup` to run a process that survives after the SSH session is terminated.
    
6. What are the interactions between `client`, `firewall`, `EC2` and `RDS` when you send a request to your app server and get the response? You can use sequence diagram to describe it more clearly.
    
    ```mermaid
    sequenceDiagram
        participant Client
        participant Firewall
        participant EC2 (App Server)
        participant RDS (Database)
        
        Client ->> Firewall: Send API Request
        Firewall ->> EC2 (App Server): Allow Request
        EC2 (App Server) ->> RDS (Database): Perform Database Operation (Insert/Select)
        RDS (Database) -->> EC2 (App Server): Return Database Result
        EC2 (App Server) -->> Client: Send API Response
    ```
    
7. (advanced optional) What did you do to make clients access your APIs without the certain port number?
8. (advanced optional) Compare to the application server, what’s the web server?
- IP address can access to your APIs
    
    http://13.238.167.235:3000/
    
- Database host
    
    database-1.c7fz0dbnsepn.ap-southeast-2.rds.amazonaws.com
    
- Change summary
    - `app.js` Modification
    Two new APIs, the "User Sign-Up API" and the "User Query API," were added
    - Addition of `utils.js`
    This file serves as a utility module, providing functions and reusable code that are utilized within `app.js`
    - Addition of `middleware.js`
    This file serves as a middleware module, providing middleware that can be applies to routes in `app.js`
    - Addition of `.env`
    This file stores sensitive information
